replay/models/cql.py

```diff
@@ -0,0 +1,357 @@
+        alpha_learning_rate: float = 1e-4,
+        actor_optim_factory: OptimizerFactory = AdamFactory(),
+        critic_optim_factory: OptimizerFactory = AdamFactory(),
+        temp_optim_factory: OptimizerFactory = AdamFactory(),
+        alpha_optim_factory: OptimizerFactory = AdamFactory(),
+        actor_encoder_factory: EncoderArg = "default",
+        critic_encoder_factory: EncoderArg = "default",
+        q_func_factory: QFuncArg = "mean",
+        batch_size: int = 256,
+        n_frames: int = 1,
+        n_steps: int = 1,
+        gamma: float = 0.99,
+        tau: float = 0.005,
+        n_critics: int = 2,
+        initial_temperature: float = 1.0,
+        initial_alpha: float = 1.0,
+        alpha_threshold: float = 10.0,
+        conservative_weight: float = 5.0,
+        n_action_samples: int = 10,
+        soft_q_backup: bool = False,
+        use_gpu: UseGPUArg = False,
+        scaler: ScalerArg = None,
+        action_scaler: ActionScalerArg = None,
+        reward_scaler: RewardScalerArg = None,
+        **params
+    ):
+        super().__init__()
+        self.top_k = top_k
+        self.action_randomization_scale = action_randomization_scale
+        self.n_epochs = n_epochs
+        self.assert_omp_single_thread()
+
+        self.model = CQL_d3rlpy.CQL(
+            actor_learning_rate=actor_learning_rate,
+            critic_learning_rate=critic_learning_rate,
+            temp_learning_rate=temp_learning_rate,
+            alpha_learning_rate=alpha_learning_rate,
+            actor_optim_factory=actor_optim_factory,
+            critic_optim_factory=critic_optim_factory,
+            temp_optim_factory=temp_optim_factory,
+            alpha_optim_factory=alpha_optim_factory,
+            actor_encoder_factory=actor_encoder_factory,
+            critic_encoder_factory=critic_encoder_factory,
+            q_func_factory=q_func_factory,
+            batch_size=batch_size,
+            n_frames=n_frames,
+            n_steps=n_steps,
+            gamma=gamma,
+            tau=tau,
+            n_critics=n_critics,
+            initial_temperature=initial_temperature,
+            initial_alpha=initial_alpha,
+            alpha_threshold=alpha_threshold,
+            conservative_weight=conservative_weight,
+            n_action_samples=n_action_samples,
+            soft_q_backup=soft_q_backup,
+            use_gpu=use_gpu,
+            scaler=scaler,
+            action_scaler=action_scaler,
+            reward_scaler=reward_scaler,
+            **params
+        )
+
+    def _fit(
+        self,
+        log: DataFrame,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+    ) -> None:
+        train: MDPDataset = self._prepare_data(log)
+        self.model.fit(train, n_epochs=self.n_epochs)
+
+    def _prepare_data(self, log: DataFrame) -> MDPDataset:
+        user_logs = log.toPandas().sort_values(['user_idx', 'timestamp'], ascending=True)
```

Comment:
What are you sorting for? If the next action is another sorting

Reply:
Such pre-sorting gets us the naturally ordered log grouped by users, which is used further on everywhere — to make `observations`, `actions` and `terminals`. Another sorting is used only to get `top K` items for each user.